research/object_detection/core/preprocessor.py

```diff
@@ -982,20 +964,18 @@ def random_adjust_saturation(image,
         generator_func,
         preprocessor_cache.PreprocessorCache.ADJUST_SATURATION,
         preprocess_vars_cache)
-    image = tf.image.adjust_saturation(image, saturation_factor)
-    image = tf.clip_by_value(image, clip_value_min=0.0, clip_value_max=1.0)
+    image = tf.image.adjust_saturation(image / 255, saturation_factor) * 255
```

Comment:
any integer images to be in [0, MAX], representing fix point numbers in [0,1).
Got that ,thank you ,what do you mean representing fix point numbers in [0,1)? At calucate time , or encodes & decodes images with tf-record time?

Reply:
I don't know... [0,255]x3/4 seems a pretty standard, fast, space efficient and intuitive way to encode 24/32 bit images...
In any case, this would fix my issue with 0 precision, clearly the 'standard' way to use [0, 1) causes problems to plenty of people. can't you do some kind of format check? or force the image into a specific format when you load it so it's consistent all across? clearly it should've thrown an error right away when it read a 200 when it should be less than 1 to begin with. but instead it's clipped to 1 and i get a bunch of garbage that the network is supposed to make sense of.

Reply:
Sorry that is confusing. For example, int8 images are stored as [0,255]. But conceptually, they represent values in [0,1) because they are fix-point numbers with an (implicit) factor of 1/256 (see also https://en.wikipedia.org/wiki/Fixed-point_arithmetic). 

That is how convert_image_dtype sees them, and how they're treated implicitly when handed to functions which accept multiple data types. Many preprocessing functions convert to float to perform their calculations, e.g. adjust_saturation will convert to float, perform a HSV conversion, adjust S, convert back to RGB, and then convert back to the original data type (or it used to, maybe we added direct int8 support for HSV conversions, but I don't think so).

Reply:
So, the best way to deal with images is either: 

- Keep the image in int8. Be careful what you're doing and make sure that all preprocessing operations can actually work on int8 images (be careful, adjust_brightness looks like it can, but it's super-dangerous because of overflows). This is faster if you do it right, but only then.

- Immediately after reading, convert the images to float. This is generally the best options, few image augmentation pipelines are able to run well in int only.

Reply:
Thanks @martinwicke for the explanation! Yes the second way is definitely prefered.

Ideally we should convert image to [0,1) at the beginning of the pipeline, but unfortunately current flow didn't follow this. We already have an ongoing effort to consolidation our interface and this issue will be gone soon.

So for now, I suggested to have this quick fix to make this model easier for git users to use.
Basically we should clearly document that these preprocessing function takes int8(or whatever) dtype as input and then use convert_image_dtype to do the conversion (instead of directly divided by 255).

Does this sound good to you @martinwicke ?

Reply:
Totally. It's up to you. Just wanted to make sure the reasoning is clear. If you can move towards this it'll be good, but I definitely don't want to block this.